File: plugins/wise/engine/src/phases/request-review.ts
// `request-review`: attach the configured reviewers with `gh pr edit --add-reviewer`
// (request-review-auto.md §1-2). Best effort throughout: a missing PR, an unknown login or
// a `gh` failure is logged and never fails the unit. The watch phase confirms the bots.

import { errText, gh, jsonOf, ok, pass } from "./common.ts";
import type { PhaseCtx, PhaseResult, PhaseRunner } from "./common.ts";

type ReviewRequests = { reviewRequests?: unknown };

async function currentReviewers(ctx: PhaseCtx, number: number): Promise<Set<string>> {
  const r = await gh(ctx, ["pr", "view", String(number), "--json", "reviewRequests"]);
  const v = jsonOf(r) as ReviewRequests | undefined;
  const out = new Set<string>();
  if (!v || !Array.isArray(v.reviewRequests)) return out;
  for (const row of v.reviewRequests as { login?: unknown; name?: unknown }[]) {
    if (typeof row.login === "string") out.add(row.login.toLowerCase());
    if (typeof row.name === "string") out.add(row.name.toLowerCase());
  }
  return out;
}

export const requestReviewPhase: PhaseRunner = async (ctx): Promise<PhaseResult> => {
  const pr = ctx.unit.pr;
  if (!pr) {
    ctx.log("request-review: no PR recorded, skipped");
    return pass();
  }
  if (ctx.config.reviewers.length === 0) {
    ctx.log("request-review: no reviewers configured, skipped");
    return pass();
  }
  const have = await currentReviewers(ctx, pr.number);
  for (const login of ctx.config.reviewers) {
    if (have.has(login.toLowerCase())) {
      ctx.log(`request-review: ${login} already requested`);
      continue;
    }
    const r = await gh(ctx, ["pr", "edit", String(pr.number), "--add-reviewer", login]);
    ctx.log(
      ok(r)
        ? `request-review: ${login} attached`
        : `request-review: ${login} unavailable (${errText(r, 120)})`,
    );
  }
  return pass();
};

File: plugins/wise/engine/wise_engine/phases/request_review.py
from ..yaml_compat import js_string
from .common import Json, err_text, gh, json_of, ok, pass_


async def request_review_phase(ctx: Json) -> Json:
    pr = ctx["unit"].get("pr")
    if not pr:
        ctx["log"]("request-review: no PR recorded, skipped")
        return pass_()
    if not ctx["config"]["reviewers"]:
        ctx["log"]("request-review: no reviewers configured, skipped")
        return pass_()
    result = json_of(
        await gh(ctx, ["pr", "view", js_string(pr["number"]), "--json", "reviewRequests"])
    )
    have: set[str] = set()
    if isinstance(result, dict) and isinstance(result.get("reviewRequests"), list):
        for row in result["reviewRequests"]:
            if isinstance(row, dict):
                have.update(
                    row[key].lower() for key in ("login", "name") if isinstance(row.get(key), str)
                )
    for login in ctx["config"]["reviewers"]:
        if login.lower() in have:
            ctx["log"](f"request-review: {login} already requested")
            continue
        res = await gh(ctx, ["pr", "edit", js_string(pr["number"]), "--add-reviewer", login])
        ctx["log"](
            f"request-review: {login} attached"
            if ok(res)
            else f"request-review: {login} unavailable ({err_text(res, 120)})"
        )
    return pass_()
